### DISCUSSION TITLE
Trying to get 2 plugins to work together

### DISCUSSION THREAD

**Discussion by s22-tech** [YOURLS/YOURLS#3197] [Plugins]
I'm trying to get my [email-notify](https://github.com/s22-tech/Yourls-email-notify/) plugin to work with the [append-qs](https://github.com/drockney/append-qs) plugin however the two don't play well together (they work well separately, though).

The email-notify plugin appears to be triggered first, and so never sees the appended querystring.  It uses 
`yourls_add_action('pre_redirect', 's22_email_notification');`
to call the function.

append-qs is using
`yourls_add_filter('redirect_location', 'append_qs_redirect');`
to call it's function.

How can I get Yourls to call `append-qs` first and then pass it's output to `email-notify`?  I can't find documentation that states the calling order of filters and actions and how they work together.

Thanks,
Marc

**Comment by ozh** [comment]
You can pass a priority to the filter. See https://github.com/YOURLS/YOURLS/blob/8c66db072b2fd380f3bfac2835eca4a8874b5f7b/includes/functions-plugins.php#L92

**Reply by s22-tech** [reply]
OK.  I changed append-qs to use
`yourls_add_filter('redirect_location', 'append_qs_redirect', 1);`
and email-notify to use
`yourls_add_action('pre_redirect', 's22_email_notification', 2);`
but email-notify is still being called first.

I'm just guessing here, but it appears that actions get called before filters?  How do these work together?

**Comment by s22-tech** [comment]
If I change the filter in `append-qs` to an action:
`yourls_add_action('pre_redirect', 'append_qs_redirect', 1);`
it now gets called first.

However, the appended URL is not being passed to `email-notify`.

**Reply by ozh** [reply]
3 things :
- Filters and actions are called in the same order as they are encountered in the code.
- the priority parameter is only used when 2 functions are hooked to the *same* filter or action.
- what the second plugin gets as input parameters depends on what the first plugin returns ...

**Comment by s22-tech** [comment]
Great info to have.  Thanks!  But my problem right now is getting the load order correct.  I think that may be the source of my other problems.

How do you force one plugin to always load before another one?

**Reply by ozh** [reply]
The "load order" doesn't matter. The fact that includes/functions-geo.php gets loaded after includes/functions-formatting.php doesn't matter. Similarly, the fact that user/plugins/joe-cool/plugin.php gets loaded before user/plugins/zzz/plugin.php doesn't matter.

All that may matter is, if two functions are hooked to the same filter or action, which one gets triggered first? Here comes the priority parameter.

**Comment by s22-tech** [comment]
OK, that makes sense since the load order did change after I made them both use the same action as I mentioned above.  But I didn't think it was working since the output from `append-qs` is still not being passed to `email-notify`.

```
yourls_add_action('pre_redirect', 'append_qs_redirect', 1);

function append_qs_redirect($url) {
...
	return $url[0].$append_qs;
}
```


```
yourls_add_action('pre_redirect', 's22_email_notification', 2);

function s22_email_notification($args) {

	---> How do I retrieve the output of the append-qs plugin here?  <---
...

}
```

**Reply by dgw** [reply]
2 comes after 1.

https://github.com/YOURLS/YOURLS/blob/23d21fd996e771c4015f5d8b4ec1b06af122d60b/includes/functions-plugins.php#L218-L219

**Reply by ozh** [reply]
Now that you show code, I see the problem. Filters are value modifiers and they return a value. Actions are events. They don't return values. 

See here https://github.com/YOURLS/YOURLS/blob/8c66db072b2fd380f3bfac2835eca4a8874b5f7b/includes/functions.php#L217 for instance:
you don't have something like `$some_value = yourls_do_action("pre_redirect", $location, $code)`, it's just the `yourls_do_action()` part. So the return statement in your plugin function is useless.

So, 2 ways for your problem:
- either hook in a *filter*, not an action
- or use an action and store somewhere values that are changed and supposed to be used later.

**Comment by s22-tech** [comment]
I'm sorry, but I don't know what you're trying to say.  I thought I had the priorities set correctly.

**Reply by dgw** [reply]
I misunderstood the problem, sorry.

(btw, you can reply to a thread instead of posting a new top-level answer every time)

**Reply by s22-tech** [reply]
So how do I retrieve the output of the `append-qs` plugin in the `email-notify` plugin?
